accepted/00000-upgrading-bootstrap.md

```diff
@@ -0,0 +1,70 @@
+- Feature Name: Upgrading Bootstrap
+- Start Date: 2022-07-27
+
+# Summary
+[summary]: #summary
+
+Suma currently uses [Bootstrap 3](https://getbootstrap.com/docs/3.3/) as its core layout and design library. Bootstrap's styles are at the core of our theme, its components and layouts are used widely, and Bootstrap permeates every piece of UI in general. Bootstrap 3 [reached EOL in 2019](https://blog.getbootstrap.com/2019/07/24/lts-plan/) and no longer receives updates of any kind, including security updates. Bootstrap 3 has [known security vulnerabilities](https://snyk.io/test/npm/bootstrap/3.3.7) which we ship today. This RFC outlines a plan to upgrade to Bootstrap 5.  
+
+# Motivation
+[motivation]: #motivation
+
+There are two main reasons to upgrade Bootstrap: ease of maintenance, and security issues.
+
+Bootstrap 3 comes from an era where browser support was more complicated and it reflects in the stylesheets we have. The current bundled stylesheet for Suma is roughly 10k lines of CSS, which is hard to reason about and maintain to say the least. This problem isn't solely due to Bootstrap, but it is the primary source of the issue. The manner in which Suma has been styled over the years has been essentially add only: since the styles are tightly coupled yet globally shared, removing things has been hard while adding a single override for any given problem is fairly easy. This has resulted in our own themes becoming complex as well, on top of an already large Bootstrap base.  
+
+As a separate issue, Bootstrap 3 has known security vulnerabilities, such as XSS injection, for which there won't be a patch. These security issues by themselves do not pose too serious of a threat due to the nature of the product, but they might be leveraged to create a chain of attacks. Given Suma is open source and has many moving parts, this is a very realistic attack vector.  
+
+The core proposition of this RFC can be summarized in two steps:  
+
+ 1. Simplify our existing stylesheets to reduce the use of custom overrides, and try and compose views and layouts using standard Bootstrap solutions. This step is beneficial even without the Bootstrap upgrade in mind, as currently our stylesheets are simply very convoluted and hard to reason about.  
+ 2. Upgrade Bootstrap. This includes many moving parts which will be outlined below.  
+
+Coming out successfully from the other side, we would hope to have simpler stylesheets with less to maintain, and fewer security issues.  
+
+
+# Detailed design
+[design]: #detailed-design
+
+## Theme & styles
+
+Bootstrap 3 uses Less while Bootstrap 5 only offers Sass styles. This switch by itself is not very problematic, however there are additional issues that make the upgrade more difficult. Our current styles make heave use of numerous mixins from Bootstrap 3 which no longer exist in Bootstrap 5. Additionally, there are numerous breaking styling changes going from 3 to 4 and even more so moving from 4 to 5.  
```

Comment:
> This switch by itself is not very problematic

Do automated translation tools exists, or at lease references/guidelines?

Reply:
Automated tools exist to translate most of Less into Sass. There are edge cases which aren't translatable since the languages function slightly differently, but I wouldn't be too worried about those as I would expect the manual effort to be limited. The main problem I would say is the mixins, since some of them no longer exist or have changed in the newer version.

Reply:
Thanks for the information. Please add a note about usage of translating tools if that is your current best plan going forward.